### DISCUSSION TITLE
Can the route hierarchy API be improved?

### DISCUSSION THREAD

**Discussion by marbemac** [TanStack/router#646] [Ideas]
The current API requires duplicating hierarchy information in multiple spots (when constructing the route tree, and on the route itself via the `getParentRoute` property).

What happens if these don't match up? Take the example below, where the `postsIndexRoute` parent is set to `postsRoute`, but that hierarchy is not reflected in the `routeTree`.

**posts-home.tsx**

```tsx
import { Route } from '@tanstack/router'
import * as React from 'react'
import { postsRoute } from '../posts'

export const postsIndexRoute = new Route({
  // postsRoute indicated as the parent here
  getParentRoute: () => postsRoute,
  path: '/',
  component: () => <div>Select a post.</div>,
})
```

**router.tsx**

```tsx
export const routeTree = rootRoute.addChildren([
  indexRoute,
  // not nested under postsRoute, what happens?
  postsIndexRoute,
])
```

---

Is this API required to support typings, or for some other reason? Will it work with nested lazy loaded routes?

Curious, because in my (very subjective :)) opinion, the config APIs in react-router and solid-router:

1. Offer a clearer at a glance picture of the entire route/path structure (with tanstack-router atm one must open all the route files to understand the application's path hierarchy) 
2. Don't allow the end user to make this kind of mis-config with their route tree

**Comment by tannerlinsley** [comment]
The type safety requires it. I wish it didn’t, but it’s currently a limitation of the compiler that isn’t going away any time soon. The route tree hierarchy is the one that matters for both runtime and the resulting router type. gerParentRoute is a type-only getter that enables child routes to be aware of and use their parent and grandparent routes types. They are both required to achieve the level of type safety the library exhibits :)
